### PR TITLE
feat: Show Reactions in Message Details - WPB-628

### PR DESCRIPTION
**Pull request by agisilaos**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/WPB-628" title="WPB-628" target="_blank">TaskWPB-628</a>  create Message Details view with reactions list
  </td></table>
  <br />
 

----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [x] contains a reference JIRA issue number like `SQPIT-764`
  - [x] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [x] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

With this PR we introduce a new way to show who reacted and with which reaction in Message Details. For that to work, we had to change the data flow a bit and create a new struct `MessageDetailsSectionDescription` and make some adjustments.

This PR currently is targeting `receive-and-display-reactions-in-conversation-view` because #263 is not merged yet.

----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: wire-ios-data-model/sourcery/generated/AutoMockable.generated.swift [hunk 1]
```diff
@@ -14,34 +13,13 @@ import AppKit
 import LocalAuthentication
 
 @testable import WireDataModel
-
```
Comment (netbe) [inline wire-ios-data-model/sourcery/generated/AutoMockable.generated.swift]: Add this file to ignored files from swiftlint

File: wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Message Details/MessageDetailsContentViewController.swift [hunk 2]
```diff
@@ -20,6 +20,23 @@ import UIKit
...
+    let headerText: String?
+    let items: [MessageDetailsCellDescription]
+
+    init(
+        headerText: String? = nil,
+        items: [MessageDetailsCellDescription]
+    ) {
+        self.headerText = headerText
+        self.items = items
+    }
```
Comment (netbe) [inline wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Message Details/MessageDetailsContentViewController.swift]:
```suggestion
    var headerText: String? = nil
    var items: [MessageDetailsCellDescription]
```

File: wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Message Details/MessageDetailsContentViewController.swift [hunk 3]
```diff
@@ -132,25 +164,53 @@ final class MessageDetailsContentViewController: UIViewController {
...
     }
 
+    private func configureConstraints() {
+        collectionView.translatesAutoresizingMaskIntoConstraints = false
+        noResultsView.translatesAutoresizingMaskIntoConstraints = false
+        subtitleLabel.translatesAutoresizingMaskIntoConstraints = false
+
+        collectionView.fitIn(view: view)
+        subtitleBottom = subtitleLabel.bottomAnchor.constraint(equalTo: safeBottomAnchor)
+        subtitleBottom!.priority = .defaultHigh
```
Comment (netbe) [inline wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Message Details/MessageDetailsContentViewController.swift]:
```suggestion
        subtitleBottom?.priority = .defaultHigh
```

**Review by netbe — APPROVED** [review]
minor suggestions

File: wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Message Details/MessageDetailsContentViewController.swift (hunk 2 above)
Comment (agisilaos) [reply wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Message Details/MessageDetailsContentViewController.swift]: We use that initializer in a couple of places..

Comment (netbe) [reply wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Message Details/MessageDetailsContentViewController.swift]: isn't it the same, with the var you get the initializer for free

File: wire-ios-data-model/sourcery/generated/AutoMockable.generated.swift (hunk 1 above)
Comment (agisilaos) [reply wire-ios-data-model/sourcery/generated/AutoMockable.generated.swift]: Done!

File: wire-ios/Wire-iOS Tests/ReferenceImages/MessageDetailsViewControllerTests/testThatItShowsBothTabs_WhenMessageIsSeenButNotLiked.1.png
Comment (johnxnguyen) [inline wire-ios/Wire-iOS Tests/ReferenceImages/MessageDetailsViewControllerTests/testThatItShowsBothTabs_WhenMessageIsSeenButNotLiked.1.png:1]: The cell moved down, I guess because of a header?

File: wire-ios/Wire-iOS Tests/ReferenceImages/MessageDetailsViewControllerTests/testThatItShowsDifferentReactions.1.png
Comment (johnxnguyen) [inline wire-ios/Wire-iOS Tests/ReferenceImages/MessageDetailsViewControllerTests/testThatItShowsDifferentReactions.1.png:1]: The header spacing seems off.

File: wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Message Details/MessageDetailsContentViewController.swift [hunk 4]
```diff
@@ -254,28 +282,73 @@ final class MessageDetailsContentViewController: UIViewController {
...
+    func numberOfSections(in collectionView: UICollectionView) -> Int {
+        return sections.count
+    }
+
+    func collectionView(
+        _ collectionView: UICollectionView,
+        layout collectionViewLayout: UICollectionViewLayout,
+        referenceSizeForHeaderInSection section: Int
+    ) -> CGSize {
+        header.titleLabel.text = "headerText"
```
Comment (johnxnguyen) [inline wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Message Details/MessageDetailsContentViewController.swift]: to delete?

File: wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Message Details/MessageDetailsContentViewController.swift [hunk 5]
```diff
@@ -254,28 +282,73 @@ final class MessageDetailsContentViewController: UIViewController {
...
+    }
+
+    func collectionView(
+        _ collectionView: UICollectionView,
+        viewForSupplementaryElementOfKind kind: String,
+        at indexPath: IndexPath
+    ) -> UICollectionReusableView {
+        let view = collectionView.dequeueReusableSupplementaryView(
+            ofKind: UICollectionView.elementKindSectionHeader,
+            withReuseIdentifier: "SectionHeader",
```
Comment (johnxnguyen) [inline wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Message Details/MessageDetailsContentViewController.swift]: I suggest to avoid duplicating this string.

File: wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Message Details/MessageDetailsDataSource.swift [hunk 6]
```diff
@@ -138,14 +153,25 @@ final class MessageDetailsDataSource: NSObject, ZMMessageObserver, ZMUserObserve
...
             performChanges {
-                self.reactions = MessageDetailsCellDescription.makeReactionCells(message.sortedLikers)
+                self.reactions = message.usersByReaction.map { reaction, users in
+                    MessageDetailsSectionDescription(
+                        headerText: reaction.unicodeValue,
+                        items: MessageDetailsCellDescription.makeReactionCells(users)
+                    )
+                }.filter {
+                    !$0.items.isEmpty
+                }
```
Comment (johnxnguyen) [inline wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Message Details/MessageDetailsDataSource.swift]: This is not sorted. I suggest to factor this method out to avoid duplication.

File: wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Message Details/MessageDetailsDataSource.swift [hunk 7]
```diff
@@ -157,8 +183,20 @@ final class MessageDetailsDataSource: NSObject, ZMMessageObserver, ZMUserObserve
...
+                )
+            }.filter {
+                !$0.items.isEmpty
+            }
+
+            self.reactions.sort(by: { $0.items.count > $1.items.count })
+
+            self.readReceipts = [
+                MessageDetailsSectionDescription(items: MessageDetailsCellDescription.makeReceiptCell(message.sortedReadReceipts))
+            ]
```
Comment (johnxnguyen) [inline wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Message Details/MessageDetailsDataSource.swift]: This is more duplication which you can avoid if you factor it out.

File: wire-ios/Wire-iOS/Sources/UserInterface/GroupDetails/UICollectionView+GroupedSections.swift [hunk 8]
```diff
@@ -29,13 +29,18 @@ fileprivate extension UICollectionViewFlowLayout {
...
         keyboardDismissMode = .onDrag
         bounces = true
         alwaysBounceVertical = true
-        contentInset = UIEdgeInsets(top: 32, left: 0, bottom: 32, right: 0)
+        if usedInMessageDetailsVC {
+            contentInset = UIEdgeInsets(top: 12, left: 0, bottom: 32, right: 0)
+        } else {
+            contentInset = UIEdgeInsets(top: 32, left: 0, bottom: 32, right: 0)
+        }
+
```
Comment (johnxnguyen) [inline wire-ios/Wire-iOS/Sources/UserInterface/GroupDetails/UICollectionView+GroupedSections.swift]: Why is this needed?

File: wire-ios/Wire-iOS Tests/ReferenceImages/MessageDetailsViewControllerTests/testThatItShowsBothTabs_WhenMessageIsSeenButNotLiked.1.png
Comment (agisilaos) [reply wire-ios/Wire-iOS Tests/ReferenceImages/MessageDetailsViewControllerTests/testThatItShowsBothTabs_WhenMessageIsSeenButNotLiked.1.png:1]: Yes

File: wire-ios/Wire-iOS Tests/ReferenceImages/MessageDetailsViewControllerTests/testThatItShowsDifferentReactions.1.png
Comment (agisilaos) [reply wire-ios/Wire-iOS Tests/ReferenceImages/MessageDetailsViewControllerTests/testThatItShowsDifferentReactions.1.png:1]: It's because we introduced the header

File: wire-ios/Wire-iOS/Sources/UserInterface/GroupDetails/UICollectionView+GroupedSections.swift (hunk 8 above)
Comment (agisilaos) [reply wire-ios/Wire-iOS/Sources/UserInterface/GroupDetails/UICollectionView+GroupedSections.swift]: This is needed so we change the top inset for our header in MessageDetails to match the design specs. I wanted to avoid changing it in all places and that's why I introduced that property.

File: wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Message Details/MessageDetailsContentViewController.swift (hunk 5 above)
Comment (agisilaos) [reply wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Message Details/MessageDetailsContentViewController.swift]: What do you mean?

File: wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Message Details/MessageDetailsContentViewController.swift (hunk 4 above)
Comment (agisilaos) [reply wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Message Details/MessageDetailsContentViewController.swift]: If we remove that, then we have no header text for our section header. What I can do is to leave it as an empty string!? Do you have any better alternatives?

File: wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Message Details/MessageDetailsDataSource.swift (hunk 6 above)
Comment (agisilaos) [reply wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Message Details/MessageDetailsDataSource.swift]: Done!

File: wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Message Details/MessageDetailsDataSource.swift (hunk 7 above)
Comment (agisilaos) [reply wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Message Details/MessageDetailsDataSource.swift]: Done!

File: wire-ios/Wire-iOS Tests/ReferenceImages/MessageDetailsViewControllerTests/testThatItShowsBothTabs_WhenMessageIsSeenButNotLiked.1.png
Comment (johnxnguyen) [reply wire-ios/Wire-iOS Tests/ReferenceImages/MessageDetailsViewControllerTests/testThatItShowsBothTabs_WhenMessageIsSeenButNotLiked.1.png:1]: Can we fix it?

File: wire-ios/Wire-iOS Tests/ReferenceImages/MessageDetailsViewControllerTests/testThatItShowsDifferentReactions.1.png
Comment (johnxnguyen) [reply wire-ios/Wire-iOS Tests/ReferenceImages/MessageDetailsViewControllerTests/testThatItShowsDifferentReactions.1.png:1]: Can we fix it?

File: wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Message Details/MessageDetailsContentViewController.swift (hunk 4 above)
Comment (johnxnguyen) [reply wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Message Details/MessageDetailsContentViewController.swift]: But what is the actual value? And why is it set here?

File: wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Message Details/MessageDetailsContentViewController.swift (hunk 5 above)
Comment (johnxnguyen) [reply wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Message Details/MessageDetailsContentViewController.swift]: `"SectionHeader"` occurs in this file multiple times. If you change on occurrence and not the other then it will cause a crash. Therefore it's best to store the value in a constant or compute the value so if you change it in one place it will change in all places.

File: wire-ios/Wire-iOS/Sources/UserInterface/GroupDetails/UICollectionView+GroupedSections.swift (hunk 8 above)
Comment (johnxnguyen) [reply wire-ios/Wire-iOS/Sources/UserInterface/GroupDetails/UICollectionView+GroupedSections.swift]: Can it be set externally? Something like:

```swift
let collectionView = ...
collectionView.contentInset = ...
```

File: wire-ios/Wire-iOS Tests/ReferenceImages/MessageDetailsViewControllerTests/testThatItShowsBothTabs_WhenMessageIsSeenButNotLiked.1.png
Comment (agisilaos) [reply wire-ios/Wire-iOS Tests/ReferenceImages/MessageDetailsViewControllerTests/testThatItShowsBothTabs_WhenMessageIsSeenButNotLiked.1.png:1]: It's fixed with my latest commits.

File: wire-ios/Wire-iOS Tests/ReferenceImages/MessageDetailsViewControllerTests/testThatItShowsDifferentReactions.1.png
Comment (agisilaos) [reply wire-ios/Wire-iOS Tests/ReferenceImages/MessageDetailsViewControllerTests/testThatItShowsDifferentReactions.1.png:1]: It's fixed with my latest commits.

File: wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Message Details/MessageDetailsContentViewController.swift (hunk 4 above)
Comment (agisilaos) [reply wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Message Details/MessageDetailsContentViewController.swift]: That's also fixed with my latest commits.

File: wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Message Details/MessageDetailsContentViewController.swift [hunk 9]
```diff
@@ -28,6 +37,8 @@ final class MessageDetailsContentViewController: UIViewController {
 
     typealias MessageDetails = L10n.Localizable.MessageDetails
 
+    private var header = SectionHeader(frame: .zero)
+
```
Comment (johnxnguyen) [inline wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Message Details/MessageDetailsContentViewController.swift]: You don't need this anymore

Comment (agisilaos) [reply wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Message Details/MessageDetailsContentViewController.swift]: Done!

File: wire-ios/Wire-iOS/Sources/UserInterface/GroupDetails/UICollectionView+GroupedSections.swift (hunk 8 above)
Comment (agisilaos) [reply wire-ios/Wire-iOS/Sources/UserInterface/GroupDetails/UICollectionView+GroupedSections.swift]: Done!

File: wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Message Details/MessageDetailsContentViewController.swift (hunk 5 above)
Comment (agisilaos) [reply wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Message Details/MessageDetailsContentViewController.swift]: Done!

**Review by johnxnguyen — APPROVED** [review]
